google-cloud-spanner/lib/google/cloud/spanner/batch_write_results.rb

```diff
@@ -0,0 +1,142 @@
+# limitations under the License.
+
+require "google/cloud/errors"
+require "google/cloud/spanner/convert"
+
+module Google
+  module Cloud
+    module Spanner
+      ##
+      # Results of a batch write.
+      #
+      # This is a stream of {BatchWriteResults::BatchResult} objects, each of
+      # which represents a set of mutation groups applied together.
+      #
+      # Use the Ruby Enumerable module to iterate over the results.
+      #
+      class BatchWriteResults
+        ##
+        # Result of a set of mutation groups applied together.
+        #
+        class BatchResult
+          # @private
+          def initialize grpc
+            @grpc = grpc
+          end
+
+          ##
+          # The indexes of the mutation groups applied together.
+          #
+          # @return [::Array<::Integer>]
+          #
+          def indexes
+            @grpc.indexes.to_a
+          end
+
+          ##
+          # The result of this set of mutation groups.
+          #
+          # @return [::Google::Rpc::Status]
+          #
+          def status
+            @grpc.status
+          end
+
+          ##
+          # Whether these mutation groups were successful.
+          #
+          # @return [::Boolean]
+          #
+          def ok?
+            status.code.zero?
+          end
+
+          ##
+          # Whether these mutation groups were unsuccessful.
+          #
+          # @return [::Boolean]
+          #
+          def error?
+            !ok?
+          end
+
+          ##
+          # The timestamp of the commit.
+          #
+          # @return [::Time]
+          #
+          def commit_timestamp
+            Convert.timestamp_to_time @grpc.commit_timestamp
+          end
+        end
+
+        # @private
+        def initialize enum
+          @enumerable = enum
+        end
+
+        ##
+        # Iterate over the results.
+        #
+        # @yield [::Google::Cloud::Spanner::BatchWriteResults::BatchResult]
+        #
+        def each &block
+          if defined? @results
+            @results.each(&block)
+          else
+            results = []
+            @enumerable.each do |grpc|
+              result = BatchResult.new grpc
+              results << result
+              yield result
+            end
+            @results = results
+          end
+        rescue GRPC::BadStatus => e
+          raise Google::Cloud::Error.from_error(e)
+        end
+
+        include Enumerable
+
+        ##
+        # Whether all mutation groups were successful.
+        #
+        # @return [::Boolean]
+        #
+        def ok?
+          all?(&:ok?)
+        end
+
+        ##
+        # Whether at least one mutation group encountered an error.
+        #
+        # @return [::Boolean]
+        #
+        def error?
+          !ok?
+        end
+
+        ##
+        # A list of the indexes of successful mutation groups.
+        #
+        # @return [::Array<::Integer>]
+        #
+        def ok_indexes
```

Comment:
`successful_mutation_indexes` (or indices)? Unless this name was already predetermined.

Reply:
You're right, that's an awkward name. Changed to `successful_indexes`.